### PR TITLE
[ZeRO][BE] Clean up ZeRO tests

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#73842 [ZeRO][BE] Clean up ZeRO tests**

**Overview**
This cleans up the `ZeroRedundancyOptimizer` tests. I apologize for strong formatting changes mixed in with actually-beneficial changes. It was convenient to unify the formatting while doing a deep comb through the full test file.

The main non-formatting changes include:
- Using `@parametrize` instead of manually including `for` loops over possible argument values
- Removing the `DEVICE` global variable, which was used only for the `TestZeroRedundancyOptimizerSingleRank` tests, in favor of consistent usage of `self.device` in both `TestZeroRedundancyOptimizerSingleRank` and `TestZeroRedundancyOptimizerDistributed`
- Moving `assert ... == ...` to `self.assertEqual(..., ...)` when the assert is part of the test's correctness
- Removing the `if self.rank >= self.world_size or (torch.cuda.is_available() and torch.cuda.device_count() < 2):` conditional guards in favor of `@common_distributed.skip_if_no_gpu` for `TestZeroRedundancyOptimizerDistributed`
    - For `TestZeroRedundancyOptimizerDistributed`, `self.device` is `torch.device(self.rank)` if CUDA is available, while `self.world_size` is at least 2, even if `torch.cuda.device_count() == 1`.
    - The problematic case is exactly when `torch.cuda.device_count() == 1` but `self.world_size == 2` since then calling `self.device` on rank 1 will error. The existing conditional guard prevented this case for some tests, but it was not used consistently (e.g. `test_multiple_groups()`), which is most likely the reason for the hangs and resulting test flakiness. (From my experience landing the recent ZeRO constructor changes, the Windows environment uses a world size of 2 but only has 1 device available.)
    - A more robust solution is to always use the `skip_if_no_gpu` decorator as long as the test uses `self.device` and CUDA is available. This is in line with the recommended SPSD usage of ZeRO.
- Renaming `test_multiple_groups()` to `test_nondefault_process_group()`
    - The existing `test_multiple_groups()` was slightly misnamed. Also, it is only nontrivial for a world size of (at least) 4 since it tests using a process group including only even ranks. It was marked as flaky on Windows, and I believe this is because of the world size and `torch.cuda.device_count()` mismatch. Now, the test only uses GPU if there are enough available and falls back to CPU otherwise, which is safe since the test uses Gloo backend.
    - There was also a duplicated section, which I was unsure how to non-naively de-duplicate. The top half and bottom half are identical even though they claim to target fitting into the broadcast bucket and not fitting into the broadcast bucket:
https://github.com/pytorch/pytorch/blob/1d497114e7e13822df403082975d4b212d836207/test/distributed/optim/test_zero_redundancy_optimizer.py#L658-L684
- Changing `_test_zero_model_parallel()` to not use CPU
    - This is my own fault, having introduced this inefficiency last summer. It makes more sense to simply designate one of the two GPUs for a process to be its default device rather than routing through CPU.

**Questions**
- How might we limit the runs for `test_ddp_zero_overlap()`? Because it parameterizes over many values, it contributes significantly to the time-to-signal. However, it is an experimental feature, so it is not critical that the tests run every time.

Differential Revision: [D34675709](https://our.internmc.facebook.com/intern/diff/D34675709)

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34675709).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34675709).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34675709).

**Comment by awgu** [comment]
Some internal tests are failing due to the test name parser (interacting with `@parametrize`). I should have the fix ready locally, but I will wait to push to avoid re-triggering the CI tests.

The takeaway is that the Github CI is now passing, even for the previously broken tests on Windows (e.g. https://github.com/pytorch/pytorch/issues/66059).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34675709).

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

File: test/distributed/optim/test_zero_redundancy_optimizer.py [hunk 1]
```diff
@@ -1163,120 +1302,68 @@ def _test_ddp_zero_overlap(
...
-        Check that overlapping DDP with ZeRO using ``hook_with_zero_step()``
-        achieves parity with DDP using a local optimizer when running on GPU.
-
-        NOTE: The test is skipped if using Windows since functional optimizers
-        are not currently supported.
+    @parametrize(
+        "use_gpu",
+        [True],
+        # Add `False` once the Gloo sync issue causing hangs is fixed
+        # See: https://github.com/pytorch/pytorch/issues/62300
```
Comment (rohan-varma) [inline test/distributed/optim/test_zero_redundancy_optimizer.py:1322]: wondering if we have plans to fix this issue in gloo? 

cc @awgu @mrshenli @cbalioglu

Comment (cbalioglu) [reply test/distributed/optim/test_zero_redundancy_optimizer.py:1322]: Assigned #62300 to myself. I plan to work on it as part of the overhaul of our process group APIs.

**Pull request closed without merging**